proposals/0000-decl-invis-binders.rst

```diff
@@ -0,0 +1,440 @@
+
+  type C :: forall i. (i -> i -> i) -> Constraint
+  class C (a :: i -> i -> i) where
+    p :: P a i
+
+This works, but now the ``i -> i -> i`` part is repeated twice, resulting in
+undesirable code duplication. One might attempt to deduplicate by removing the
+inline kind annotation on ``a``::
+
+  type C :: forall i. (i -> i -> i) -> Constraint
+  class C a where
+    p :: P a i
+
+Alas, now GHC will interpret this code very differently! The ``i`` in the
+standalone kind signature for ``C`` and the ``i`` in the class method ``p`` are
+no longer in any way related, and the type of ``p`` changes as follows::
+
+  p :: forall i (a :: i -> i -> i). C a => P a i                -- before
+  p :: forall i (a :: i -> i -> i) k (i1 :: k). C a => P a i1   -- after
+
+This wouldn't happen if ``i`` from the standalone kind signature scoped over
+the class body, as one might expect with ``ScopedTypeVariables`` enabled.
+However, this is simply not the case: ``StandaloneKindSignatures`` and
+``ScopedTypeVariables`` do not interact. One might argue that this is a reason
+to make them interact in the expected way, but this proposal offers no
+judgement in this regard. Instead, it offers an alternative that sidesteps the
+issue entirely::
+
+  type C :: forall i. (i -> i -> i) -> Constraint
+  class C @i a where
+    p :: P a i
+
+By writing ``@i``, we bind the type variable, making it scope over the class
+body; at the same time, we avoid repeating ``i -> i -> i``.
+
+Argument 4: Quantification in Type Synonyms
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Consider::
+
+  type T1 = 'Nothing :: Maybe a
+  type T2 = 'Just ('Nothing :: Maybe a)
+
+``T1`` is currently legal, yielding ``T :: forall a. Maybe a``. The general
+rule is that the free variables of a *top-level* kind annotation on the RHS are
+brought into scope implicitly, and will be quantified in the final kind of the
+type constructor.
+
+In constrast, ``T2`` is currently illegal, because the kind annotation is not
+at the top level.
+
+We propose to drop this exotic form of implicit quantification from the
+language. Both ``T1`` and ``T2`` would become illegal, but with ``@k``-binders
+the programmer can rewrite them as follows::
+
+  type T1 @a = 'Nothing :: Maybe a
+  type T2 @a = 'Just ('Nothing :: Maybe a)
+
+This way all the variables occurring on the RHS are bound on the LHS.
+We exploit the new syntax to allow a nice, simple, uniform scoping rule.
+To cite the User's Guide, "The reason for this exception [the strange, ad-hoc
+rule about top-level kind annotations] is that there may be no other way to
+bind k".
+
+The same rule applies to type family instances::
+
+   type family F a :: k
+
+   type instance F Int = Any :: j -> j
+
+(where ``Any :: forall p. p``). This type instance is currently legal, but
+under this proposal ``j`` would not be in scope. You would have to write::
+
+   type instance F @(j->j) Int = Any :: j -> j
+
+In type instances, this form is already allowed today.
```

Comment:
What does this sentence mean?  This *is* a `type instance` declaration!  Maybe you mean class `instance`?   Give an example

Reply:
Yes, it is a type instance declaration, and it is already allowed